File: src/engine/rules.ts
/**
 * กติกาการเดินและการกิน — pure functions ทั้งหมด
 *
 * หลักการ:
 *  - หมากทุกตัวเป็นกลาง ไม่มีเจ้าของ ใครถึงเทิร์นเลือกหมากตัวใดก็ได้
 *  - กิน = กระโดดข้ามหมาก 1 ตัวไปยังช่องถัดไปที่ต้องว่าง (8 ทิศ)
 *  - กินต่อเนื่องได้ และเปลี่ยนทิศระหว่าง chain ได้
 *  - เมื่อเลือกหมากแล้ว ต้องเดิน chain ที่กินได้ "มากที่สุด" ของหมากตัวนั้น
 *  - ถ้าหมากที่เลือกกินไม่ได้เลย จึงเดินปกติ 1 ช่องได้ (8 ทิศ) แล้วจบเทิร์น
 */

import { type Board, type Dir, DIRS, shift } from "./board.js";

export interface CaptureStep {
  /** ช่องต้นทางของก้าวนี้ */
  from: number;
  /** ช่องของหมากที่ถูกข้าม (จะถูกกินออกจากกระดาน) */
  over: number;
  /** ช่องปลายทาง (ต้องว่าง) */
  to: number;
  /** piece id ของหมากที่ถูกกิน */
  capturedPiece: number;
  dir: Dir;
}

/**
 * เพดานจำนวน node ของการค้นหา chain กันกรณีกระดานที่แตกแขนงมหาศาล
 * ในทางปฏิบัติของกระดาน 8x8 แทบไม่มีทางแตะเพดานนี้
 */
const SEARCH_NODE_BUDGET = 2_000_000;

/** ก้าวการกินที่เป็นไปได้ทั้งหมดจากช่อง `from` (1 ก้าว ไม่ดู chain) */
export function captureStepsFrom(board: Board, from: number): CaptureStep[] {
  const steps: CaptureStep[] = [];
  if (board[from] === null) return steps;
  for (const dir of DIRS) {
    const over = shift(from, dir, 1);
    if (over < 0) continue;
    const to = shift(from, dir, 2);
    if (to < 0) continue;
    const captured = board[over];
    if (captured === null) continue;
    if (board[to] !== null) continue;
    steps.push({ from, over, to, capturedPiece: captured, dir });
  }
  return steps;
}

/** ค้นหาความยาว chain สูงสุดจากช่อง `from` (mutate/undo บน board ที่ให้มา) */
function searchMaxChain(board: Board, from: number, budget: { n: number }): number {
  let best = 0;
  const piece = board[from];
  if (piece === null) return 0;
  for (const dir of DIRS) {
    if (budget.n <= 0) break;
    const over = shift(from, dir, 1);
    if (over < 0) continue;
    const to = shift(from, dir, 2);
    if (to < 0) continue;
    const captured = board[over];
    if (captured === null || board[to] !== null) continue;

    budget.n--;
    board[from] = null;
    board[over] = null;
    board[to] = piece;
    const depth = 1 + searchMaxChain(board, to, budget);
    board[to] = null;
    board[over] = captured;
    board[from] = piece;

    if (depth > best) best = depth;
  }
  return best;
}

export interface CaptureAnalysis {
  /** จำนวนหมากสูงสุดที่หมากตัวนี้กินได้ใน 1 เทิร์น (0 = กินไม่ได้) */
  max: number;
  /** ก้าวแรกที่ "ยังคงพา chain ไปถึงค่าสูงสุด" ได้เท่านั้น */
  steps: CaptureStep[];
}

/**
 * วิเคราะห์การกินของหมากที่ช่อง `from`
 * คืนค่า max = ความยาว chain สูงสุด และ steps = ก้าวถัดไปที่ถูกกติกา
 * (กติกา Maximum Capture: ต้องเลือกก้าวที่ยังไปถึง max ได้เท่านั้น)
 */
export function analyzeCaptures(board: Board, from: number): CaptureAnalysis {
  const first = captureStepsFrom(board, from);
  if (first.length === 0) return { max: 0, steps: [] };

  const work = board.slice();
  const piece = work[from]!;
  const budget = { n: SEARCH_NODE_BUDGET };
  const depths: number[] = [];
  let max = 0;

  for (const step of first) {
    work[from] = null;
    work[step.over] = null;
    work[step.to] = piece;
    const depth = 1 + searchMaxChain(work, step.to, budget);
    work[step.to] = null;
    work[step.over] = step.capturedPiece;
    work[from] = piece;

    depths.push(depth);
    if (depth > max) max = depth;
  }

  return { max, steps: first.filter((_, i) => depths[i] === max) };
}

/** ช่องปลายทางของการเดินปกติ: 1 ช่อง 8 ทิศ ไปยังช่องที่ว่าง */
export function moveDestinations(board: Board, from: number): number[] {
  const dests: number[] = [];
  if (board[from] === null) return dests;
  for (const dir of DIRS) {
    const to = shift(from, dir, 1);
    if (to >= 0 && board[to] === null) dests.push(to);
  }
  return dests;
}

export type PieceOptions =
  | { kind: "capture"; max: number; steps: CaptureStep[] }
  | { kind: "move"; destinations: number[] }
  | { kind: "none" };

/**
 * ตัวเลือกทั้งหมดของหมากที่ช่อง `from`
 * ถ้ากินได้ ต้องกิน (และต้องกินให้ได้มากที่สุด) — เดินปกติไม่ได้
 */
export function pieceOptions(board: Board, from: number): PieceOptions {
  if (from < 0 || from >= board.length || board[from] === null) return { kind: "none" };
  const captures = analyzeCaptures(board, from);
  if (captures.max > 0) return { kind: "capture", max: captures.max, steps: captures.steps };
  const destinations = moveDestinations(board, from);
  if (destinations.length > 0) return { kind: "move", destinations };
  return { kind: "none" };
}

/** หมากตัวนี้มีอะไรให้ทำไหม (กินได้ หรืออย่างน้อยขยับได้) */
export function pieceHasAction(board: Board, from: number): boolean {
  if (board[from] === null) return false;
  if (moveDestinations(board, from).length > 0) return true;
  return captureStepsFrom(board, from).length > 0;
}

/** ทุกช่องที่ผู้เล่นหยิบมาเล่นได้ในเทิร์นนี้ */
export function selectableSquares(board: Board): number[] {
  const out: number[] = [];
  for (let i = 0; i < board.length; i++) {
    if (board[i] !== null && pieceHasAction(board, i)) out.push(i);
  }
  return out;
}

/** ยังมีการเล่นที่ถูกต้องเหลืออยู่ไหม (ใช้เป็นเงื่อนไขจบเกม) */
export function hasAnyLegalAction(board: Board): boolean {
  for (let i = 0; i < board.length; i++) {
    if (board[i] !== null && pieceHasAction(board, i)) return true;
  }
  return false;
}

File: apps/api/src/makthai/games/mak_si_kradan/rules.py
"""กติกาการกินและการเดิน — ฟังก์ชันล้วน ไม่มี state ของตัวเอง

หลักการ:
  * หมากทุกตัวเป็นของกลาง ใครถึงเทิร์นเลือกตัวไหนก็ได้
  * กิน = กระโดดข้ามหมาก 1 ตัวไปยังช่องถัดไปที่ต้องว่าง ได้ทั้ง 8 ทิศ
  * กินต่อเนื่องได้และเปลี่ยนทิศระหว่างทางได้
  * ถ้าหมากที่เลือกกินไม่ได้ ก็ขยับ 1 ช่องใน 8 ทิศแทน
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from makthai.games.mak_si_kradan.board import DIRECTIONS, Board, Direction, shift

#: เพดานจำนวน node ของการค้นหา chain กันกระดานที่แตกแขนงมหาศาล
#: บนกระดาน 8x8 แทบไม่มีทางแตะเพดานนี้
SEARCH_NODE_BUDGET = 2_000_000


@dataclass(frozen=True, slots=True)
class CaptureStep:
    """การกระโดดกินหนึ่งก้าว"""

    origin: int
    over: int
    landing: int
    captured_piece: int
    direction: Direction


@dataclass(frozen=True, slots=True)
class CaptureAnalysis:
    #: จำนวนหมากสูงสุดที่หมากตัวนี้กินได้ใน 1 เทิร์น (0 = กินไม่ได้)
    max_captures: int
    #: ก้าวแรกที่ยังพา chain ไปถึงค่าสูงสุดได้เท่านั้น
    steps: tuple[CaptureStep, ...]


def capture_steps_from(board: Board, origin: int) -> list[CaptureStep]:
    """ก้าวการกินที่เป็นไปได้ทั้งหมดจากช่องหนึ่ง โดยยังไม่ดูว่า chain จะยาวแค่ไหน"""
    steps: list[CaptureStep] = []
    if board[origin] is None:
        return steps
    for direction in DIRECTIONS:
        over = shift(origin, direction, 1)
        if over < 0:
            continue
        landing = shift(origin, direction, 2)
        if landing < 0:
            continue
        captured = board[over]
        if captured is None or board[landing] is not None:
            continue
        steps.append(CaptureStep(origin, over, landing, captured, direction))
    return steps


def _search_max_chain(board: Board, origin: int, budget: list[int]) -> int:
    """ความยาว chain สูงสุดจากช่องหนึ่ง — แก้กระดานแล้วคืนค่าเดิมเพื่อไม่ต้อง copy"""
    piece = board[origin]
    if piece is None:
        return 0

    best = 0
    for direction in DIRECTIONS:
        if budget[0] <= 0:
            break
        over = shift(origin, direction, 1)
        if over < 0:
            continue
        landing = shift(origin, direction, 2)
        if landing < 0:
            continue
        captured = board[over]
        if captured is None or board[landing] is not None:
            continue

        budget[0] -= 1
        board[origin] = None
        board[over] = None
        board[landing] = piece
        depth = 1 + _search_max_chain(board, landing, budget)
        board[landing] = None
        board[over] = captured
        board[origin] = piece

        best = max(best, depth)
    return best


def analyze_captures(board: Board, origin: int) -> CaptureAnalysis:
    """หา chain ที่ยาวที่สุดของหมากตัวหนึ่ง พร้อมก้าวแรกที่ยังไปถึงความยาวนั้นได้

    กติกา Maximum Capture บังคับว่าต้องเดินก้าวที่ยังพาไปถึงค่าสูงสุดเท่านั้น
    """
    first = capture_steps_from(board, origin)
    if not first:
        return CaptureAnalysis(0, ())

    work = list(board)
    piece = work[origin]
    budget = [SEARCH_NODE_BUDGET]
    depths: list[int] = []

    for step in first:
        work[origin] = None
        work[step.over] = None
        work[step.landing] = piece
        depths.append(1 + _search_max_chain(work, step.landing, budget))
        work[step.landing] = None
        work[step.over] = step.captured_piece
        work[origin] = piece

    best = max(depths)
    return CaptureAnalysis(best, tuple(s for s, d in zip(first, depths, strict=True) if d == best))


def move_destinations(board: Board, origin: int) -> list[int]:
    """ช่องปลายทางของการเดินปกติ — 1 ช่อง 8 ทิศ ไปยังช่องที่ว่าง"""
    if board[origin] is None:
        return []
    destinations: list[int] = []
    for direction in DIRECTIONS:
        landing = shift(origin, direction, 1)
        if landing >= 0 and board[landing] is None:
            destinations.append(landing)
    return destinations


@dataclass(frozen=True, slots=True)
class PieceOptions:
    kind: Literal["capture", "move", "none"]
    max_captures: int = 0
    steps: tuple[CaptureStep, ...] = ()
    destinations: tuple[int, ...] = ()


def piece_options(board: Board, origin: int) -> PieceOptions:
    """สิ่งที่หมากตัวนี้ทำได้ — กินได้ก็ถือว่าเป็นตัวกิน ไม่งั้นก็เป็นตัวเดิน"""
    if not (0 <= origin < len(board)) or board[origin] is None:
        return PieceOptions("none")
    analysis = analyze_captures(board, origin)
    if analysis.max_captures > 0:
        return PieceOptions("capture", analysis.max_captures, analysis.steps)
    destinations = move_destinations(board, origin)
    if destinations:
        return PieceOptions("move", destinations=tuple(destinations))
    return PieceOptions("none")


def piece_has_action(board: Board, origin: int) -> bool:
    if board[origin] is None:
        return False
    return bool(move_destinations(board, origin)) or bool(capture_steps_from(board, origin))


def selectable_squares(board: Board) -> list[int]:
    """ทุกช่องที่หยิบมาเล่นได้ในเทิร์นนี้"""
    return [i for i in range(len(board)) if board[i] is not None and piece_has_action(board, i)]


def has_any_legal_action(board: Board) -> bool:
    return any(board[i] is not None and piece_has_action(board, i) for i in range(len(board)))
